### PR TITLE
MED-19

### PR DESCRIPTION
**Pull request by Damageddream**
Added Appointment logic

### CODE REVIEW

File: src/main/java/com/damageddream/medicalclinic/repository/AppointmentRepository.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+import java.time.LocalDateTime;
+import java.util.List;
+
+public interface AppointmentRepository extends JpaRepository<Appointment, Long> {
+
+    @Query("SELECT a FROM Appointment a WHERE a.doctor.id = :doctorId AND " +
+            "((:newAppointmentStart >= a.appointmentStart AND :newAppointmentStart < a.appointmentEnd) OR " +
+            "(:newAppointmentEnd > a.appointmentStart AND :newAppointmentEnd <= a.appointmentEnd) OR " +
+            "(:newAppointmentStart <= a.appointmentStart AND :newAppointmentEnd >= a.appointmentEnd))")
```
Comment (Kustlik) [inline src/main/java/com/damageddream/medicalclinic/repository/AppointmentRepository.java:16]: Query could be greatly simplified
`@Query("SELECT a FROM Appointment a WHERE :doctorId = a.doctor.id AND a.appointmentStart  <=:newAppointmentStart AND a.appointmentEnd >=:newAppointmentStart")`

File: src/main/java/com/damageddream/medicalclinic/rest/AppointmentRestController.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+@RequestMapping("/appointments")
+public class AppointmentRestController {
+    private final AppointmentService appointmentService;
+    @PostMapping("/doctor/{id}")
+    public AppointmentDTO addAppointment(@PathVariable Long id, @RequestBody Appointment appointment) {
+        return appointmentService.addAppointment(id, appointment);
+    }
+
+    @GetMapping("/doctor/{id}/free")
+    public List<AppointmentDTO> getFreeAppointments(@PathVariable Long id){
```
Comment (Kustlik) [inline src/main/java/com/damageddream/medicalclinic/rest/AppointmentRestController.java:22]: I would move this method to AppointmentRestController.java

File: src/main/java/com/damageddream/medicalclinic/entity/Doctor.java [hunk 3]
```diff
@@ -44,4 +46,7 @@ public boolean equals(Object o) {
     public int hashCode() {
         return getClass().hashCode();
     }
-    }
+
+}
+
+
```
Comment (konradgab) [inline src/main/java/com/damageddream/medicalclinic/entity/Doctor.java:52]: nadmiarowa pusta linia

File: src/main/java/com/damageddream/medicalclinic/rest/PatientRestController.java [hunk 4]
```diff
@@ -47,4 +45,6 @@ public PatientDTO updatePassword(@PathVariable String patientEmail, @RequestBody
     public PatientDTO deletePatient(@PathVariable String patientEmail) {
         return patientService.delete(patientEmail);
     }
+
+
```
Comment (konradgab) [inline src/main/java/com/damageddream/medicalclinic/rest/PatientRestController.java:49]: nadmiarowa pusta linia

File: src/main/java/com/damageddream/medicalclinic/rest/advice/ServiceExceptionHandler.java [hunk 5]
```diff
@@ -90,6 +90,23 @@ public ResponseEntity<ErrorResponse> facilityNotFoundExceptionResponse(
 
         return ResponseEntity.status(HttpStatus.NOT_FOUND).body(errorResponse);
     }
```
Comment (konradgab) [inline src/main/java/com/damageddream/medicalclinic/rest/advice/ServiceExceptionHandler.java:92]: brak odstepu miedzy metodami

File: src/main/java/com/damageddream/medicalclinic/service/AppointmentServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+@Service
+@RequiredArgsConstructor
+public class AppointmentServiceImpl implements AppointmentService {
+
+    private final AppointmentMapper appointmentMapper;
+    private final PatientRepository patientRepository;
+    private final DoctorRepository doctorRepository;
+    private final AppointmentRepository appointmentRepository;
+
+    private final DataValidator dataValidator;
```
Comment (konradgab) [inline src/main/java/com/damageddream/medicalclinic/service/AppointmentServiceImpl.java:33]: czemu w oddzielnej linii i nie ma linii odstępu od metod

File: src/main/java/com/damageddream/medicalclinic/service/AppointmentServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        appointment.setDoctor(doctor);
+        appointmentRepository.save(appointment);
+
+        return appointmentMapper.toDTO(appointment);
+    }
+
+    @Override
+    public List<AppointmentDTO> getFreeAppointmentsByDoctor(Long doctorId) {
+        List<Appointment> appointments = appointmentRepository.findFreeAppointmentsByDoctor(doctorId);
+        if(appointments.isEmpty()){
```
Comment (konradgab) [inline src/main/java/com/damageddream/medicalclinic/service/AppointmentServiceImpl.java:57]: ctrl alt l

File: src/main/java/com/damageddream/medicalclinic/service/AppointmentServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    @Override
+    @Transactional
+    public AppointmentDTO makeAnAppointment(Long patientId, GetIdCommand appointmentIdCommand) {
+        Long appointmentId = appointmentIdCommand.getEntityId();
+        Patient patient = patientRepository.findById(patientId)
+                .orElseThrow(()->new PatientNotFoundException("Patient not found"));
+        Appointment appointment = appointmentRepository.findById(appointmentId).orElseThrow(
+                ()-> new AppointmentNotFoundException("There is no such appointment")
+        );
+        if(appointment.getPatient() != null){
```
Comment (konradgab) [inline src/main/java/com/damageddream/medicalclinic/service/AppointmentServiceImpl.java:72]: ctrl alt l

File: src/main/java/com/damageddream/medicalclinic/service/AppointmentServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+            throw  new AppointmentNotFoundException("That appointment is no longer free");
+        }
+        appointment.setPatient(patient);
+        return appointmentMapper.toDTO(appointment);
+    }
+
+    @Override
+    public List<AppointmentDTO> getPatientsAppointments(Long id) {
+        Patient patient = patientRepository.findById(id)
+                .orElseThrow(()->new PatientNotFoundException("Patient not found"));
```
Comment (konradgab) [inline src/main/java/com/damageddream/medicalclinic/service/AppointmentServiceImpl.java:82]: ctrl alt l

File: src/main/java/com/damageddream/medicalclinic/service/AppointmentServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+        List<Appointment> appointments = appointmentRepository.findAppointmentsByPatient(id);
+        if(appointments.isEmpty()){
```
Comment (konradgab) [inline src/main/java/com/damageddream/medicalclinic/service/AppointmentServiceImpl.java:85]: ctrl alt l

File: src/main/java/com/damageddream/medicalclinic/service/DoctorServiceImpl.java [hunk 11]
```diff
@@ -68,4 +69,6 @@ public DoctorDTO addFacilityToDoctor(Long doctorId, GetIdCommand entityId) {
         doctorRepository.save(doctor);
         return doctorMapper.toDTO(doctor);
     }
+
+
```
Comment (konradgab) [inline src/main/java/com/damageddream/medicalclinic/service/DoctorServiceImpl.java:73]: nadmiarowa pusta linia

File: src/main/java/com/damageddream/medicalclinic/service/AppointmentService.java [hunk 12]
```diff
@@ -0,0 +1,15 @@
+package com.damageddream.medicalclinic.service;
+
+import com.damageddream.medicalclinic.dto.AppointmentDTO;
+import com.damageddream.medicalclinic.dto.GetIdCommand;
+import com.damageddream.medicalclinic.entity.Appointment;
+
+import java.util.List;
+
+public interface AppointmentService {
+    AppointmentDTO makeAnAppointment(Long patientId, GetIdCommand appointmentId);
```
Comment (Kustlik) [inline src/main/java/com/damageddream/medicalclinic/service/AppointmentService.java:10]: I would change name to maybe something like IdentifierDTO

File: src/main/java/com/damageddream/medicalclinic/validation/DataValidator.java [hunk 13]
```diff
@@ -45,4 +55,29 @@ private void validateIfAllNewPatientFieldsAreFilled(NewPatientDTO newPatientData
             throw new PatientFieldNullException("All fields of new Patient must be completed");
         }
     }
+
+    private void validateIfdateIsNotPast(LocalDateTime startDate){
+        LocalDateTime now = LocalDateTime.now().plusHours(2);
+        if(startDate.isBefore(now)){
+            throw new InvalidDateTimeException("You have to choose date after "+now.toString());
```
Comment (konradgab) [inline src/main/java/com/damageddream/medicalclinic/validation/DataValidator.java:62]: ctrl alt l

File: src/main/java/com/damageddream/medicalclinic/validation/DataValidator.java [hunk 14]
```diff
@@ -45,4 +55,29 @@ private void validateIfAllNewPatientFieldsAreFilled(NewPatientDTO newPatientData
...
+
+    private void validateIfEndNotBeforeStart(LocalDateTime startDate, LocalDateTime endDate) {
+        if(endDate.isBefore(startDate) || endDate.isEqual(startDate)){
+            throw new InvalidDateTimeException("End date cannot be equal or before start date");
+        }
+    }
+
+    private void validateIfAppointmentAtLeast30min(LocalDateTime startDate, LocalDateTime endDate) {
+        if(endDate.isBefore(startDate.plusMinutes(30))){
+            throw new InvalidDateTimeException("Appointment duration must be at least 30 minutes");
```
Comment (Kustlik) [inline src/main/java/com/damageddream/medicalclinic/validation/DataValidator.java:74]: I don't think this method is needed, that wasn't our requirement if I remember ritght. @konradgab is that right?

File: src/main/java/com/damageddream/medicalclinic/service/AppointmentService.java (hunk 12 above)
Comment (Damageddream) [reply src/main/java/com/damageddream/medicalclinic/service/AppointmentService.java:10]: Agree to disagree

File: src/main/java/com/damageddream/medicalclinic/dto/mapper/DoctorMapper.java [hunk 15]
```diff
@@ -4,12 +4,15 @@
 import com.damageddream.medicalclinic.dto.NewDoctorDTO;
 import com.damageddream.medicalclinic.entity.Doctor;
 import org.mapstruct.Mapper;
-
-
+import org.mapstruct.Mapping;
+import org.mapstruct.MappingTarget;
```
Comment (konradgab) [inline src/main/java/com/damageddream/medicalclinic/dto/mapper/DoctorMapper.java:10]: nadmiarowa pusta linia

File: src/main/java/com/damageddream/medicalclinic/dto/mapper/DoctorMapper.java [hunk 16]
```diff
@@ -4,12 +4,15 @@
...
 
 
 @Mapper(componentModel = "spring")
 public interface DoctorMapper {
     DoctorDTO toDTO(Doctor doctor);
+
     Doctor fromDTO(NewDoctorDTO newDoctorDTO);
```
Comment (konradgab) [inline src/main/java/com/damageddream/medicalclinic/dto/mapper/DoctorMapper.java:15]: jak robimy odstepy to robimy odstepy jak nie robimy to nie robimy

File: src/main/java/com/damageddream/medicalclinic/rest/AppointmentRestController.java [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/appointments")
+public class AppointmentRestController {
+    private final AppointmentService appointmentService;
+
+    @PostMapping("/doctor/{id}")
+    @ResponseStatus(HttpStatus.CREATED)
+    public AppointmentDTO addAppointment(@PathVariable Long id, @RequestBody Appointment appointment) {
```
Comment (konradgab) [inline src/main/java/com/damageddream/medicalclinic/rest/AppointmentRestController.java:21]: POST appointments wywalmy ta scieke z doktorem i do Appointment dodajmy pole doctorId

File: src/main/java/com/damageddream/medicalclinic/service/AppointmentServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+        }
+        return appointments.stream().map(appointmentMapper::toDTO).toList();
+    }
+
+    @Override
+    @Transactional
+    public AppointmentDTO makeAnAppointment(Long patientId, GetIdCommand appointmentIdCommand) {
+        Long appointmentId = appointmentIdCommand.getEntityId();
+        Patient patient = patientRepository.findById(patientId).orElseThrow(() -> new PatientNotFoundException("Patient not found"));
+        Appointment appointment = appointmentRepository.findById(appointmentId).orElseThrow(() -> new AppointmentNotFoundException("There is no such appointment"));
```
Comment (konradgab) [inline src/main/java/com/damageddream/medicalclinic/service/AppointmentServiceImpl.java:64]: Appointment appointment = appointmentRepository.findById(appointmentId)
.orElseThrow(() -> new AppointmentNotFoundException("There is no such appointment"));

File: src/main/java/com/damageddream/medicalclinic/service/DoctorService.java [hunk 19]
```diff
@@ -9,8 +9,16 @@
...
+
     DoctorDTO save(NewDoctorDTO newDoctorDTO);
+
     DoctorDTO addFacilityToDoctor(Long doctorId, GetIdCommand entityId);
 
+    DoctorDTO deleteDoctor(Long id);
+
+    DoctorDTO update(Long id, NewDoctorDTO newDoctor);
+
+
```
Comment (konradgab) [inline src/main/java/com/damageddream/medicalclinic/service/DoctorService.java:23]: nadmiarowa pusta linia

File: src/main/java/com/damageddream/medicalclinic/service/FacilityServiceImpl.java [hunk 20]
```diff
@@ -61,12 +63,36 @@ public FacilityDTO addDoctorToFacility(Long facilityId, GetIdCommand entityId) {
...
+        facilityMapper.updateFacilityFromDTO(newFacility, facility);
+        facilityRepository.save(facility);
+        return facilityMapper.toDTO(facility);
+    }
+
+    @Override
+    @Transactional
+    public FacilityDTO deleteFacility(Long id) {
+        Facility facility = facilityRepository.findById(id)
+                .orElseThrow(() -> new FacilityNotFoundException("Facility not found"));
```
Comment (konradgab) [inline src/main/java/com/damageddream/medicalclinic/service/FacilityServiceImpl.java:89]: sprobujmy cascade type all lub Delete dac  i sprawdzic dzialanie

**Pull request merged**